control/injector/Injector.php

```diff
@@ -551,6 +551,14 @@ protected function instantiate($spec, $id=null, $type = null) {
 			$constructorParams = $spec['constructor'];
 		}
 
+		// If we're dealing with a DataObject, pass through Singleton flag as second argument
+		if ($type != 'prototype' && in_array('DataObject', ClassInfo::ancestry($class))) {
```

Comment:
A lot of binding here to DataObject. How about looking into a DataObjectFactory configured via yaml?

Reply:
I don't really know what that means. I really want to see this ancient issue fixed but I don't have a lot of time to look further in to it or solve it more elegantly unfortunately. I had thought this would be only for 3.x too but actually glancing at Injector and DataObject in SS4 this is probably still an issue and would need to be merged up.

Reply:
Kk well, fair enough. I'd probably not want this leaking up into 4.x branch though. 4 should be less concerned about singleton-specific construction.